File: FDK/SoftFX.Extended.Net/Financial/Collections/PriceEntries.cs
namespace SoftFX.Extended.Financial
{
    using System.Collections;
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// Provides access to price entries collection.
    /// </summary>
    public class PriceEntries : IEnumerable<KeyValuePair<string, PriceEntry>>
    {
        #region Construction

        /// <summary>
        /// Creates a new instance of price entires.
        /// </summary>
        public PriceEntries()
        {
            this.entries = new SortedDictionary<string, PriceEntry>();
        }

        #endregion

        #region Methods

        /// <summary>
        /// 
        /// </summary>
        /// <param name="symbol"></param>
        /// <returns></returns>
        public PriceEntry? TryGetPriceEntry(string symbol)
        {
            var result = default(PriceEntry);
            if (this.entries.TryGetValue(symbol, out result))
                return result;

            return null;
        }

        /// <summary>
        /// Gets all prices as dictionary.
        /// </summary>
        /// <returns></returns>
        public Dictionary<string, PriceEntry> ToDictionary()
        {
            var result = this.entries.ToDictionary(k => k.Key, v => v.Value);
            return result;
        }

        /// <summary>
        /// Updates bid/ask prices for a symbol.
        /// </summary>
        /// <param name="symbol">an updating symbol</param>
        /// <param name="bid">a bid price</param>
        /// <param name="ask">an ask price</param>
        public void Update(string symbol, double bid, double ask)
        {
            this.entries[symbol] = new PriceEntry(bid, ask);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="prices"></param>
        public void Update(IEnumerable<KeyValuePair<string, PriceEntry>> prices)
        {
            foreach (var element in prices)
            {
                this.entries[element.Key] = element.Value;
            }
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="symbol"></param>
        public void Remove(string symbol)
        {
            this.entries.Remove(symbol);
        }

        /// <summary>
        /// 
        /// </summary>
        public void Clear()
        {
            this.entries.Clear();
        }

        #endregion

        #region IEnumerable interface implementation

        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        public IEnumerator<KeyValuePair<string, PriceEntry>> GetEnumerator()
        {
            return this.entries.GetEnumerator();
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.entries.GetEnumerator();
        }

        #endregion

        #region Fields

        readonly SortedDictionary<string, PriceEntry> entries;

        #endregion
    }
}

File: FDK/LlCommon/PriceEntries.cpp
#include "stdafx.h"
#include "PriceEntries.h"
namespace FDK
{
	CPriceEntries::CPriceEntries()
	{
	}

	CPriceEntries::CPriceEntries(const CPriceEntries& entries)
        : m_entries(entries.m_entries)
	{
	}

	CPriceEntries& CPriceEntries::operator = (const CPriceEntries& entries)
	{
		if (this != &entries)
		{
			m_entries = entries.m_entries;
		}
		return *this;
	}

	CPriceEntries::~CPriceEntries()
	{
	}

	void CPriceEntries::Update(const char* symbol, const double bid, const double ask)
	{
		string key = symbol;
		m_entries[key] = CPriceEntry(bid, ask);
	}

	void CPriceEntries::Remove(const char* symbol)
	{
		string key = symbol;
		m_entries.erase(key);
	}

	void CPriceEntries::Clear()
	{
		m_entries.clear();
	}

	Nullable<CPriceEntry> CPriceEntries::TryGetPriceEntry(const char* symbol) const
	{
		string key = symbol;
		auto it = m_entries.find(key);
		if (m_entries.end() != it)
		{
			return it->second;
		}
		return nullptr;
	}

	const LrpMap(string, CPriceEntry)& CPriceEntries::GetEntries() const
	{
		return m_entries;
	}
}